pkg/cmd/roachtest/tests/cluster_to_cluster.go

```diff
@@ -650,17 +655,69 @@ AS OF SYSTEM TIME '%s'`, startTime.AsOfSystemTime(), endTime.AsOfSystemTime())
 	// If the goroutine gets cancelled or fataled, return before comparing fingerprints.
 	require.NoError(rd.t, fingerPrintMonitor.WaitE())
 	if srcFingerprint != destFingerprint {
-		rd.t.L().Printf("fingerpint mismatch: conducting table level fingerprints")
-		srcTenantConn := rd.c.Conn(ctx, rd.t.L(), 1, option.TenantName(rd.setup.src.name))
-		dstTenantConn := rd.c.Conn(ctx, rd.t.L(), rd.rs.srcNodes+1, option.TenantName(rd.setup.dst.name))
-		require.NoError(rd.t, replicationutils.InvestigateFingerprints(ctx, srcTenantConn, dstTenantConn,
-			startTime,
-			endTime))
-		rd.t.L().Printf("fingerprints by table seem to match")
+		rd.onFingerprintMismatch(ctx, startTime, endTime)
 	}
 	require.Equal(rd.t, srcFingerprint, destFingerprint)
 }
 
+// onFingerprintMismatch runs table level fingerprints and backs up both tenants.
+func (rd *replicationDriver) onFingerprintMismatch(
+	ctx context.Context, startTime, endTime hlc.Timestamp,
+) {
+	rd.t.L().Printf("conducting table level fingerprints")
+	srcTenantConn := rd.c.Conn(ctx, rd.t.L(), 1, option.TenantName(rd.setup.src.name))
+	dstTenantConn := rd.c.Conn(ctx, rd.t.L(), rd.rs.srcNodes+1, option.TenantName(rd.setup.dst.name))
+	fingerprintBisectErr := replicationutils.InvestigateFingerprints(ctx, srcTenantConn, dstTenantConn,
+		startTime,
+		endTime)
+	// Before failing on this error, back up the source and destination tenants.
+	if fingerprintBisectErr != nil {
+		rd.t.L().Printf("fingerprint bisect error", fingerprintBisectErr)
+	} else {
+		rd.t.L().Printf("table level fingerprints seem to match")
+	}
+	rd.t.L().Printf("backing up src and destination tenants")
+	fingerPrintMonitor := rd.newMonitor(ctx)
+
+	fingerPrintMonitor.Go(func(ctx context.Context) error {
+		return rd.backupAfterFingerprintMismatch(ctx, srcTenantConn, rd.setup.src.name, startTime, endTime)
+	})
+	fingerPrintMonitor.Go(func(ctx context.Context) error {
+		return rd.backupAfterFingerprintMismatch(ctx, dstTenantConn, rd.setup.dst.name, startTime, endTime)
+	})
+	fingerprintMonitorError := fingerPrintMonitor.WaitE()
+	require.NoError(rd.t, errors.CombineErrors(fingerprintBisectErr, fingerprintMonitorError))
+}
+
+// backupAfterFingerprintMismatch runs two backups on the provided tenant: a
+// full backup AOST the provided startTime, and an incremental backup AOST the
+// provided endTime. Both backups run with revision history.
+func (rd *replicationDriver) backupAfterFingerprintMismatch(
+	ctx context.Context, conn *gosql.DB, tenantName string, startTime, endTime hlc.Timestamp,
+) error {
+	if rd.c.IsLocal() {
+		rd.t.L().Printf("skip taking backups of tenants on local roachtest run")
+		return nil
+	}
+	prefix := "gs"
+	if rd.c.Spec().Cloud == spec.AWS {
+		prefix = "s3"
+	}
+	collection := fmt.Sprintf("%s://%s/c2c-fingerprint-mismatch/%s/%s/%s?AUTH=implicit", prefix, testutils.BackupTestingBucket(), rd.rs.name, rd.c.Name(), tenantName)
```

Comment:
If we end up worried about the space these backups take up in the "sometimesFingerprint" case, then one thing we could do is pass in an alternate backup location such that we could backup to a bucket with some aggressive cleanup attached to it.

Reply:
heh, funny you ask: the [gcs bucket](https://console.cloud.google.com/storage/browser/cockroachdb-backup-testing;tab=lifecycle?forceOnBucketsSortingFiltering=true&project=cockroach-shared&supportedpurview=project&prefix=&forceOnObjectsSortingFiltering=false) has a ttl of 1 day, potentially too low for L2 to quickly transfer the backup to a bucket with a looser retention policy. The [s3 bucket ](https://s3.console.aws.amazon.com/s3/buckets/cockroachdb-backup-testing?region=us-east-1&tab=metrics)on the other hand has no ttl policy. I will start a thread with test eng, and potentially take action in a follow up pr.